src/persist-client/src/read.rs

```diff
@@ -62,6 +62,126 @@ impl ReaderId {
     }
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(bound(
+    serialize = "T: Timestamp + Codec64",
+    deserialize = "T: Timestamp + Codec64"
+))]
+#[serde(
+    into = "SerdeReadEnrichedHollowBatch",
+    from = "SerdeReadEnrichedHollowBatch"
+)]
+/// A `HollowBatch` plus the metadata necessary to apply the appropriate read
+/// style's semantics (i.e. snapshotting or listening).
+pub struct ReaderEnrichedHollowBatch<T> {
+    pub(crate) shard_id: ShardId,
+    pub(crate) as_of: Antichain<T>,
```

Comment:
the filtering that we need to do is dependent on what type of batch this is, so I think we should be modeling that explicitly instead of relying on the Some-or-None-ness of frontier

concretely: pull the as_of+frontier into something like an `EnrichedHollowBatchType` enum with one variant for snapshot and one for listen (the corresponding proto should be some sort of proto union). there would be an as_of field on both and only listen would have frontier

Reply:
Yep-this is a TODO elsewhere (probably should have been a TODO here)